zmq/connections.nim

```diff
@@ -0,0 +1,199 @@
+import bindings
+# Unofficial easier-for-Nim API
+
+#[
+  Types definition
+]#
+type
+  EZmq* = object of IOError ## exception that is raised if something fails
+  TConnection* {.pure, final.} = object ## a connection
+    c*: PContext                        ## the embedded context
+    s*: PSocket                         ## the embedded socket
+    ownctx: bool                        ## Does the Connection own the context ?
+    sockaddr: string                    ## Address of the underlying socket
+
+#[
+  Error handler
+]#
+proc zmqError*() {.noinline, noreturn.} =
+  ## raises EZmq with error message from `zmq.strerror`.
+  var e: ref EZmq
+  new(e)
+  e.msg = $strerror(errno())
+  raise e
+
+#[
+  get/set socket options
+  Declare socket options first because it's used in =destroy hooks
+]#
+# Some option take cint, int64 or uint64
+proc setsockopt_impl[T: SomeOrdinal](s: PSocket, option: TSockOptions, optval: T) =
+  var val: T = optval
+  if setsockopt(s, option, addr(val), sizeof(val)) != 0:
+    zmqError()
+# Some option take cstring
+proc setsockopt_impl(s: PSocket, option: TSockOptions, optval: string) =
+  var val: string = optval
+  if setsockopt(s, option, cstring(val), val.len) != 0:
+    zmqError()
+
+# some sockopt returns integer values
+proc getsockopt_impl[T: SomeOrdinal](s: PSocket, option: TSockOptions, optval: var T) =
+  var optval_len: int = sizeof(optval)
+
+  if bindings.getsockopt(s, option, addr(optval), addr(optval_len)) != 0:
+    zmqError()
+
+# Some sockopt returns a string
+proc getsockopt_impl(s: PSocket, option: TSockOptions, optval: var string) =
+  var optval_len: int = optval.len
+
+  if bindings.getsockopt(s, option, cstring(optval), addr(optval_len)) != 0:
+    zmqError()
+
+#[
+  Public set/get sockopt function on PSocket / TConnection
+]#
+proc setsockopt*[T: SomeOrdinal|string](s: PSocket, option: TSockOptions, optval: T) =
+  setsockopt_impl[T](s, option, optval)
+
+proc setsockopt*[T: SomeOrdinal|string](c: TConnection, option: TSockOptions, optval: T) =
+  setsockopt[T](c.s, option, optval)
+
+proc getsockopt*[T: SomeOrdinal|string](s: PSocket, option: TSockOptions): T =
+  var optval: T
+  getsockopt_impl(s, option, optval)
+  optval
+
+proc getsockopt*[T: SomeOrdinal|string](c: TConnection, option: TSockOptions): T =
+  getsockopt[T](c.s, option)
+
+
+#[
+  Destructor
+]#
+
+when defined(gcDestructors):
+  proc close*(c: TConnection)
+  proc `=destroy`(x: var TConnection) =
+    echo "=destroy"
+    # Set linger to 0 to properly drop message
+    setsockopt(x, LINGER, 0.cint)
+    close(x)
```

Comment:
Instead of closing in the destructor which then cannot really report errors back, assert in the destructor that `close` was called.

Reply:
I raise an exception if destructor is called without previous manual close. Was that what you hadi n mind ?

Reply:
That is what I had in mind.